### PR TITLE
Distinct position check

### PR DESCRIPTION
**Pull request by softarius**
### Summary of Changes
Method  _getListCount of class BaseDatabaseModel select query with keyword "distinct"

### Testing Instructions
- insert into any table repeated values
- create a model extends ListModel
- in method getListQuery  write SQL query for getting "dictionary" like "select distinct field1 from table1"
- write view and layout with pagination for getting a table this dictionary 
I wrote a component (only PostgreSQL supported) for demonstrating (attached).
Use URL `joomlasite/index.php?option=com_person&view=names`

[com_person.zip](https://github.com/joomla/joomla-cms/files/3260519/com_person.zip)

### Expected result
Correct pagination
espected

### Actual result
Pagination with the wrong page counting, because distinct practically ignored
actual

### Documentation Changes Required
Nothing

### CODE REVIEW

**Comment by ghost** [comment]
If this is a new Feature please rebaso on J4 as J3 is only Bug-Fixing.

**Comment by alikon** [comment]
@franz-wohlkoenig 
sounds like a bug to me

File: libraries/src/MVC/Model/BaseDatabaseModel.php [hunk 1]
```diff
@@ -330,7 +330,8 @@ protected function _getListCount($query)
 			&& $query->group === null
 			&& $query->union === null
 			&& $query->unionAll === null
-			&& $query->having === null)
+			&& $query->having === null
+			&& stripos((string) $query, 'distinct') === false)
```
Comment (zero-24) [inline libraries/src/MVC/Model/BaseDatabaseModel.php:334]:
```suggestion
			&& stripos((string) strtolower($query), 'distinct') === false)
```

hmm should't we do a strtolower($query) here too as i could also write that key word in all upper case.

**Review by zero-24 — CHANGES_REQUESTED** [review]

File: libraries/src/MVC/Model/BaseDatabaseModel.php (hunk 1 above)
Comment (HLeithner) [reply libraries/src/MVC/Model/BaseDatabaseModel.php:334]: stripos is already case insensitive no need to change this.

**Review by zero-24 — APPROVED** [review]

**Comment by HLeithner** [comment]
It's a bug but should test all core components list views.

**Comment by SharkyKZ** [comment]
Did anyone read the note?

https://github.com/joomla/joomla-cms/blob/7bc411dc2e99ce15713a31b8e6f1d7ba64d229c7/libraries/src/MVC/Model/BaseDatabaseModel.php#L314-L317

**Comment by softarius** [comment]
> Did anyone read the note?
> 
> https://github.com/joomla/joomla-cms/blob/7bc411dc2e99ce15713a31b8e6f1d7ba64d229c7/libraries/src/MVC/Model/BaseDatabaseModel.php#L314-L317

I have read this note. The phrase about "must be overridden" similar to the description of the bug. Let's just fix it. 
The main idea of this fix - not use count(*) if distinct exists. It's all.

**Comment by SharkyKZ** [comment]
There is no bug if you follow the documentation. You have to override the method in your model. And you should do that not just to get the correct result count but also to improve performance. That fallback method is extremely slow and should be avoided.

Furthermore, simple string manipulation like this is going to cause the slow query to run more often than it should, i.e. whenever anything contains the string `distinct`. This could be a table name, a column, a search keyword or anything else. So at least, if maintainers don't mind, we should add distinct capability to Database API to correctly detect where distinct is actually used or not.

**Comment by richard67** [comment]
@SharkyKZ So should this PR be closed according to your previous comment?
@softarius If @SharkyKZ 's answer is yes, would you agree?

**Comment by SharkyKZ** [comment]
It's up to maintainers to decide.

**Comment by HLeithner** [comment]
I think you are right it brings more harm and the dev should do it right as in the description.

@softarius thanks anyway for your contribution.

**Pull request closed without merging**